### PR TITLE
Fixes #25396 - Add pagination row component

### PR DESCRIPTION
**Pull request by xprazak2**
Additional changes might be needed so that katello can use this, therefore marked as wip for now. Needed by https://github.com/theforeman/foreman_ansible/pull/217

### CODE REVIEW

**Comment by ares** [comment]
I'm just curious, do you expect this to be useful in other plugins or core? If that was the case, it would better fit core I think

Edit: please ignore me, I missed the repo url on phone and saw ansible in comment 0 :-)

File: webpack/assets/javascripts/react_app/components/common/PaginationRow/PaginationRow.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { Paginator } from 'patternfly-react';
+import { isEqual } from 'lodash';
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/PaginationRow/PaginationRow.js]: are you using this function?

File: webpack/assets/javascripts/react_app/components/common/PaginationRow/PaginationRow.test.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      perPage: 5,
+      perPageOptions: [5, 10, 25],
+    },
+    itemCount: 52,
+    viewType: 'list',
+  });
+
+  describe('rendering', () => {
+    it('renders correctly', () => {
+      const component = shallow(<PaginationRow {...getBaseProps()} />);
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/PaginationRow/PaginationRow.test.js]: you can use `testComponentSnapshotsWithFixtures` from `testHelpers.js`

File: webpack/assets/javascripts/react_app/components/common/PaginationRow/PaginationRow.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  ...Paginator.propTypes,
+  /** page and per-page selection callback */
+  onChange: PropTypes.func,
+  pagination: PropTypes.shape({
+    /** the current page */
+    page: PropTypes.number,
+    /** the current per page setting */
+    perPage: PropTypes.number,
+    /** per page options */
+    perPageOptions: PropTypes.array,
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/common/PaginationRow/PaginationRow.js]: please use `arrayOf` or `shape`, much more informative.

**Review by amirfefer — CHANGES_REQUESTED** [review]
We have already a [pagination component](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Pagination/Pagination.js), which looks almost identical, if you'd like to add some behaviors or do some refactoring, why not consume/change the existing component?

**Comment by xprazak2** [comment]
Updated, katello worked for me when I was testing.

@amirfefer: I know there is Pagination component. It relies on being used directly by `mount_react_component` helper and uses `window.location.href` to change pages, so the two are not really compatible. Changing it would mean choosing different implementation based on props, effectively something like this:

```js
const PaginationWrapper = (props) => {
  if (props.data) {
    return <Pagination ...props />
  }

  return <PaginationRow ...props />
}

```

I decided not to go that way, but let me know if you would like me to mash the 2 components together in some way.

**Comment by ohadlevy** [comment]
@glekner can you please have a look? thanks.

**Comment by glekner** [comment]
I'm against creating a new component.
Rendering a different component based on props seems like the right choice to me.

Is there another functionality that you can't implement with the current pagination besides changing pages?

If not, We can create a default behavior for the erb case : 
```
Pagination.propTypes = {
...props,
navigateTo: PropTypes.func
}

Pagination.defaultProps = {
navigateTo: (url) => window.Turbolinks.visit(url.toString())
}
```

**Comment by xprazak2** [comment]
I merged the two components together as suggested.

File: webpack/assets/javascripts/react_app/components/Pagination/Pagination.js [hunk 4]
```diff
@@ -5,35 +5,108 @@ import { Paginator } from 'patternfly-react';
...
+}
+
+
+const initPagination = (paginationFromProps) => {
+  const pagination = paginationFromProps || {};
+
+  const defaultPagination = {
+    page: 1,
+    perPage: 20,
+    perPageOptions: [5, 10, 15, 20, 25, 50],
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/Pagination/Pagination.js]: does it make sense to have all of these options as default propTypes?

**Comment by glekner** [comment]
I still see a lot of duplicated code, it doesn't matter that its in the same file.

I wrote this [gist](https://gist.github.com/glekner/3aa1b32d2414685f6f50275976171062), it should work for both cases with defaultProps. The only required prop is `data`:

```
  data: Proptypes.shape({
    viewType: Proptypes.string,
    perPageOptions: Proptypes.arrayOf(Proptypes.number),
    itemCount: Proptypes.number,
    perPage: Proptypes.number,
  }).isRequired,
```

File: webpack/assets/javascripts/react_app/components/Pagination/Pagination.js (hunk 4 above)
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/Pagination/Pagination.js]: The defaults are shallow-merged, so having it in defaults does not really work if `{ page: 1, perPage: 10 }` is passed in.

**Comment by xprazak2** [comment]
@glekner, I made changes based on the gist you suggested. Setting the erb behaviour as default does not work as I need the defaults from Paginator.

**Comment by glekner** [comment]
> @glekner, I made changes based on the gist you suggested. Setting the erb behaviour as default does not work as I need the defaults from Paginator.

I'm not quite sure i understand, the whole point is you overriding the default behaviour.
you need to create a stateful component that will render my gist.
The changes you did are great and basically fine by me but still a lot of code can be deleted.
What default props are your relying on from the Paginator?

In order for the component to work for erb all you need to do is put the state inside a `data` object as a prop.

I can try to do it next week, but if you have time:
it should be something like this:

```
class StatefulWrapper {
this.state = {
page,
perPage,
...etc
}

setPerPage ()
setPage ()

render() {

<Pagination
data = { { page: this.state.page, perPage: this.state.perPage ... } }
setPerPage={this.setPerPage}
setPage={this.setPage}
...etc />
}
```

**Comment by xprazak2** [comment]
@glekner, I see what you mean now, I added `PaginationRow` as a wrapper component. But wasn't creating a new component something you wanted to [avoid](https://github.com/theforeman/foreman/pull/6208#issuecomment-436677052)?

File: webpack/assets/javascripts/react_app/components/Pagination/PaginationRow.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      page: 1,
+      perPage: 20,
+      perPageOptions: [5, 10, 15, 20, 25, 50],
+    };
+    return { ...defaultPagination, ...(paginationFromProps || {}) };
+  };
+
+  return (
+    <Pagination
+      data={{}}
```
Comment (glekner) [inline webpack/assets/javascripts/react_app/components/Pagination/PaginationRow.js]: Why is data empty? how is `viewType, itemCount` being passed to `Pagination`? looks like its getting passed by `...otherProps` which is not ideal, you should just create a data object in your wrapper for more detailed propTypes 

```suggestion
      data={{ itemCount }}
```

**Review by glekner — COMMENTED** [review]
I wasn't specific enough, Creating a **Wrapper** component is great IMO. Great work!

small comment, and maybe we should think about renaming `PaginationRow` so people know its a wrapper for js

File: webpack/assets/javascripts/react_app/components/Pagination/Pagination.js [hunk 6]
```diff
@@ -2,28 +2,51 @@ import React from 'react';
...
+        isEmpty(pagination)
+          ? {
+              page: urlPage || 1,
+              perPage: urlPerPage || data.perPage,
+              perPageOptions: data.perPageOptions,
+            }
+          : pagination
+      }
+      viewType={data.viewType}
+      itemCount={data.itemCount}
```
Comment (glekner) [inline webpack/assets/javascripts/react_app/components/Pagination/Pagination.js:42]: the compnent is expecting the itemCount from data, aswell as viewType but it looks like youre not inputing a viewType so its using the default one

**Comment by xprazak2** [comment]
I moved the `itemCount` and `viewType` into `data` and renamed `PaginationRow` to `PaginationWrapper`

**Comment by glekner** [comment]
LGTM! @xprazak2 , make sure Travis tests pass.

**Comment by xprazak2** [comment]
Travis is green

**Comment by xprazak2** [comment]
@tstrachota, could you take a look please?

**Review by tstrachota — COMMENTED** [review]
It looks good and works great in Katello.

Would you mind adding a storybook for the `PaginationWrapper` component? It's highly reusable component and I find it important to have some usage examples.

**Comment by xprazak2** [comment]
I added a storybook example

**Comment by tstrachota** [comment]
[test foreman]

**Review by tstrachota — APPROVED** [review]
Thank you @xprazak2

**Pull request merged**